src/Ocelot/Infrastructure/DesignPatterns/Retry.cs

```diff
@@ -0,0 +1,111 @@
+﻿using Ocelot.Logging;
+
+namespace Ocelot.Infrastructure.DesignPatterns;
+
+/// <summary>
+/// Basic <seealso href="https://www.bing.com/search?q=Retry+pattern">Retry pattern</seealso> for stabilizing integrated services.
+/// </summary>
+/// <remarks>Docs:
+/// <list type="bullet">
+/// <item><see href="https://learn.microsoft.com/en-us/azure/architecture/patterns/retry">Microsoft Learn | Retry pattern</see></item>
+/// </list>
+/// </remarks>
+public static class Retry
+{
+    public const int DefaultRetryTimes = 3;
+    public const int DefaultWaitTimeMilliseconds = 25;
+
+    /// <summary>
+    /// Retry a synchronous operation when an exception occurs or predicate is true, then delay and retry again.
+    /// </summary>
+    /// <typeparam name="TResult">Type of the result of the sync operation.</typeparam>
+    /// <param name="operation">Required Func-delegate of the operation.</param>
+    /// <param name="predicate">Predicate to check, optionally.</param>
+    /// <param name="retryTimes">Number of retries.</param>
+    /// <param name="waitTime">Waiting time in milliseconds.</param>
+    /// <param name="logger">Concrete logger from upper context.</param>
+    /// <returns>A <typeparamref name="TResult"/> value as the result of the sync operation.</returns>
+    public static TResult Operation<TResult>(
+        Func<TResult> operation,
+        Predicate<TResult> predicate = null,
+        int retryTimes = DefaultRetryTimes, int waitTime = DefaultWaitTimeMilliseconds,
+        IOcelotLogger logger = null)
+    {
+        for (int i = 0; i < retryTimes - 1; i++)
+        {
+            TResult result = default;
```

Comment:
The default is not needed here... By the way, sometimes, I'm wondering why we are reeinventing the wheel everywhere when we could simply use polly for the retry...

Reply:
We cannot adopt Polly for several reasons:
- Polly has [PredicateBuilder](https://www.pollydocs.org/api/Polly.PredicateBuilder-1.html) feature but its another reinvention of the wheel for Polly fans who can't provide lite solutions. Also docs says in remarks: "Please be aware of the performance penalty if you register too many predicates with this builder. In such case, it's better to create your own predicate manually as a delegate." So I did manually! I believe my implementation has better performance.
- It is a large and complex library with numerous settings, which is unnecessary for simple tasks.
- Using Polly alongside other libraries can be risky for developers. We need to adhere to a single integration package, namely `KubeClient`, and referencing `Polly` is huge design mistake in my opinion.
Debugging the attached Polly pipeline will be a developer's nightmare for us! New `Retry` methods will be debugged easily!

In contrast, this `Retry` offers a lightweight implementation, giving us more control over the code compared to Polly, which has a substantial amount of partially "closed" code.
Sorry, I can't add `Polly` reference to [Ocelot.Provider.Kubernetes.csproj](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot.Provider.Kubernetes/Ocelot.Provider.Kubernetes.csproj#L31-L36) package! And I wondered that you are asking me to do that 🤯

Reply:
>The default is not needed here... 

OK It is minor issue, we could still assign. But seems compiler does the same even without [default literal](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/default#default-literal)
Fair enough!

Reply:
> It's just that we have several code blocks for retry and for timeouts. Probably using polly isn't the right strategy, but you got what I meant...

I understand your points; however, [my counterarguments](https://github.com/ThreeMammals/Ocelot/pull/2111#discussion_r1705420376) to Polly can be found above in the discussion.

Tiny issue with `default` is fixed, see https://github.com/ThreeMammals/Ocelot/pull/2111/commits/c4c7ae0ff2a891d64b4f81b6a6d051cc15e71211#diff-01ceec0d1ccf6008635b11e7d2c5a6d2ef2654eaba66371ffc488d7f1b0c7563  

I hope this thread can be closed.